controllers/datafoundation_client.go

```diff
@@ -0,0 +1,241 @@
+
+	isValid := true
+	onboardingTicket, found := offering.Spec.Config["onboardingTicket"]
+	if !found {
+		r.Log.Error(
+			fmt.Errorf("missing field: onboardingTicket"),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	} else if onboardingTicket == "" {
+		r.Log.Error(
+			fmt.Errorf("onboardingTicket is empty"),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	}
+
+	dfStorageProviderEndpoint, found := offering.Spec.Config["dataFoundationProviderEndpoint"]
+	if !found {
+		r.Log.Error(
+			fmt.Errorf("missing field: dataFoundationProviderEndpoint"),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	} else if dfStorageProviderEndpoint == "" {
+		r.Log.Error(
+			fmt.Errorf("dataFoundationProviderEndpoint is empty"),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	}
+	dfStorageProviderEndpointAsURL, err := url.ParseRequestURI(dfStorageProviderEndpoint)
+	if err != nil {
+		r.Log.Error(
+			fmt.Errorf("unable to parse Data Foundation Provider Endpoint url"),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	}
+	if dfStorageProviderEndpointAsURL.Host == "" {
+		r.Log.Error(
+			fmt.Errorf("cannot retrieve host from %s, might be a malformed or partial URL", dfStorageProviderEndpoint),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	}
+
+	if !isValid {
+		r.Log.Info("parsing ManagedFusionOffering Data Foundation Client spec failed")
+		return fmt.Errorf("invalid ManagedFusionOffering Data Foundation Client spec")
+	}
+	r.Log.Info("parsing ManagedFusionOffering Data Foundation Client spec completed successfuly")
+
+	r.dataFoundationClientSpec = dataFoundationClientSpec{
+		onboardingTicket:          onboardingTicket,
+		dfStorageProviderEndpoint: dfStorageProviderEndpoint,
+	}
+	return nil
+}
+
+func (r *dataFoundationClientReconciler) reconcilePhases() (ctrl.Result, error) {
+	if !r.offering.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	} else {
+		if err := r.reconcileStorageClient(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileCSV(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileDefaultBlockStorageClassClaim(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileDefaultFSStorageClassClaim(); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+func (r *dataFoundationClientReconciler) reconcileStorageClient() error {
+	r.Log.Info("Reconciling StorageClient")
+
+	_, err := r.CreateOrUpdate(&r.storageClient, func() error {
+		if err := r.own(&r.storageClient, true); err != nil {
+			return err
+		}
```

Comment:
Please add whatever check makes sense for the values before setting them on the spec.

Reply:
updated